Add AVM fritzbox smarthome component

hthiery opened the pull request:
## Description:

AVM fritzbox component that supports AVM DECT300 and Comet DECT thermostates in the climate platform and the AVM DECT200 in the switch platform.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation:** home-assistant/home-assistant.github.io#4076

## Example entry for `configuration.yaml`:
```yaml

fritzbox:
  devices:
    - host: FRITZBOX-IP
      username: YOUR-USERNAME
      password: YOUR-PASSWORD
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.


[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

hthiery committed:
initial commit

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

tinloaf commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,138 @@
+    ATTR_ID, ATTR_MANUFACTURER, ATTR_PRODUCTNAME, DOMAIN)
+from homeassistant.components.climate import (ClimateDevice, STATE_ECO)
+from homeassistant.const import (PRECISION_HALVES)
+from homeassistant.const import (TEMP_CELSIUS, ATTR_TEMPERATURE)
```
Why write this twice?

tinloaf commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,138 @@
+
+    def __init__(self, hass, device):
+        self._device = device
+        self._actual_temperature = None
```
This is usually called "current" temperature throughout HA. Please rename to avoid confusion.

tinloaf commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,138 @@
+class FritzhomeThermostat(ClimateDevice):
+    """The thermostat class for Fritzhome."""
+
+    OPERATION_LIST = [STATE_ECO, STATE_COMFORT]
```
Why do you define this as a member? I would define this with the other constants, where you defined `STATE_COMFORT`

tinloaf commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,138 @@
+
+    @property
+    def current_temperature(self):
+        """Can not report temperature, so return target_temperature."""
```
That comment seems to be wrong.

tinloaf commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,138 @@
+        """Return the temperature we try to reach."""
+        return self._target_temperature
+
+    def set_temperature(self, **kwargs):
```
`set_temperature` can also be called with an operation, to set operation and temperature in one go. See https://home-assistant.io/components/climate/#service-climateset_temperature 

You should handle this here.

tinloaf commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,138 @@
+        return self.OPERATION_LIST
+
+    @property
+    def min_temp(self):
```
I think this should return the minimum temperature that this device can be set to. That is probably not the "eco" temperature?

hthiery replied:
There is no programmatic way to get the min/max values. Only static min/max values can be set. Is that a way to solve this?

tinloaf replied:
If you don't implement it, the default is 7°C / 35°C. I think most platforms just use these values, i.e., don't implement `min_temp()` / `max_temp()`.

tinloaf commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,138 @@
+        return self._eco_temperature
+
+    @property
+    def max_temp(self):
```
See min_temp()

tinloaf commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,138 @@
+            self._target_temperature = self._device.target_temperature
+            self._comfort_temperature = self._device.comfort_temperature
+            self._eco_temperature = self._device.eco_temperature
+        except Exception as exc:
```
I'm not sure if you shouldn't just keep the old values. Occasional update failures might happen, and setting everything to `None` could confuse the front end, automations etc.

tinloaf commented on homeassistant/components/switch/fritzhome.py:
```diff
@@ -0,0 +1,93 @@
+
+
+class FritzhomeSwitch(SwitchDevice):
+    """The thermostat class for Fritzhome."""
```
thermostat -> switch

hthiery committed:
fix failed flake8 tests

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
add fritzhome files to .coveragerc

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
fix wrong module import

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
remove too general exception

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
incorporate review comments

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
remove blank line

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
fix wrong import

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
fix issue with operations

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

balloob commented on homeassistant/components/switch/fritzhome.py:
```diff
@@ -0,0 +1,90 @@
+        from pyfritzhome import InvalidError
+
+        try:
+            state = self._device.get_switch_state()
```
We should never do I/O inside properties. Move all I/O to update method

balloob commented on homeassistant/components/switch/fritzhome.py:
```diff
@@ -0,0 +1,90 @@
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Fritzhome switch platform."""
+    if DOMAIN not in hass.data:
```
This will not happen as you have specified it in DEPENDENCIES

balloob commented on homeassistant/components/switch/fritzhome.py:
```diff
@@ -0,0 +1,90 @@
+        """Return the state attributes of the device."""
+        attr = {
+            ATTR_AIN: self._device.ain,
+            ATTR_FW_VERSION: self._device.fw_version,
```
We should never include firmware version, manufacturer or product name in the device state attributes. Please remove.

Attributes should only contain info about the state that is relevant for automations.

balloob commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,146 @@
+            ATTR_AIN: self._device.ain,
+            ATTR_FW_VERSION: self._device.fw_version,
+            ATTR_ID: self._device.id,
+            ATTR_MANUFACTURER: self._device.manufacturer,
```
Please remove these attributes.

hthiery committed:
incorporate review comments

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
remove unused attributes

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
adapt to supported_features

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery commented:
Is there anything I can do?

tinloaf commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,137 @@
+    def __init__(self, hass, device):
+        """Initialize the thermostat."""
+        self._device = device
+        self._current_temperature = self._device.actual_temperature
```
Are these values already available at this point? If so, some I/O must have happened at this point, right? It would be preferable to delay the I/O until after the platform has been set up: Set the values to `None` here, and call `add_devices` above with `True` as second parameter. This way, HA will schedule a call to your `update` method instantly.

hthiery replied:
the device with its values is already fetched in the setup of the fritzhome component and the values are valid. So I think this is ok.

tinloaf commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,137 @@
+
+    def set_temperature(self, **kwargs):
+        """Set new target temperature."""
+        if kwargs.get(ATTR_OPERATION_MODE) is not None:
```
Just a nitpick: I think the canonical way of checking this is `if ATTR_OPERATION_MODE in kwargs:`

tinloaf commented on homeassistant/components/switch/fritzhome.py:
```diff
@@ -0,0 +1,66 @@
+        if device.has_switch:
+            devices.append(FritzhomeSwitch(hass, device))
+
+    add_devices(devices)
```
You probably also want to call this with `True` as a second parameter to retrieve the switch's status right on start.

tinloaf commented on homeassistant/components/switch/fritzhome.py:
```diff
@@ -0,0 +1,66 @@
+    def __init__(self, hass, device):
+        """Initialize the switch."""
+        self._device = device
+        self._state = None
```
You never use `self._state` anywhere?

tinloaf reviewed (dismissed):
See inline comments

MartinHjelmare commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,137 @@
+"""
+import logging
+
+from homeassistant.components.fritzhome import DOMAIN
```
You should not overwrite the climate domain like this. Instead do:
```py
from homeassistant.components.fritzhome import DOMAIN as FRITZHOME_DATA
```
Then use this constant below, where needed.

MartinHjelmare commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,137 @@
+from homeassistant.components.climate import (
+    ATTR_OPERATION_MODE, ClimateDevice, STATE_ECO,
+    SUPPORT_OPERATION_MODE, SUPPORT_TARGET_TEMPERATURE)
+from homeassistant.const import (PRECISION_HALVES)
```
You don't need parenthesis if not breaking the line.

MartinHjelmare commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,137 @@
+
+SUPPORT_FLAGS = (SUPPORT_TARGET_TEMPERATURE | SUPPORT_OPERATION_MODE)
+
+OPERATION_LIST = [STATE_COMFORT, STATE_ECO, STATE_MANUAL]
```
This list should only return states that are defined in the base climate component. You're not allowed to define custom device specific states, that are user facing.

hthiery replied:
Can i add these operation modes to the climate base component? Otherwise I have a problem to map these modes to currently availables.

MartinHjelmare replied:
What do you think @tinloaf?

tinloaf replied:
I would be fine with it; however, I'd also ask you (@hthiery , I mean) not to be offended if we change this a bit when we do some refactoring in the future. Names may change because we will try to consolidate constants across platforms.

hthiery replied:
I'm fine with that until there is mode that matches the purpose. I would like to change the comfort mode in the fritzhome thermostate to use the `STATE_HEAT` instead of `STATE_COMFORT` and add `STATE_MANUAL` to the base climate component.

MartinHjelmare commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,137 @@
+class FritzhomeThermostat(ClimateDevice):
+    """The thermostat class for Fritzhome thermostates."""
+
+    def __init__(self, hass, device):
```
Don't pass in `hass`. It will be set on the entity when the entity has been added to home assistant. This means you can access `hass` with `self.hass`, after the entity has been added to home assistant.

MartinHjelmare commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,137 @@
+    @property
+    def current_operation(self):
+        """Return the current operation mode."""
+        if not self.available:
```
You can remove this. It will be handled in the entity base class.

MartinHjelmare commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,137 @@
+        if not self.available:
+            return None
+        if self._target_temperature == self._comfort_temperature:
+            return STATE_COMFORT
```
See comment above about allowed operating states.

MartinHjelmare commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,137 @@
+            return STATE_COMFORT
+        elif self._target_temperature == self._eco_temperature:
+            return STATE_ECO
+        return STATE_MANUAL
```
Same as above.

MartinHjelmare commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,137 @@
+
+    def set_operation_mode(self, operation_mode):
+        """Set new operation mode."""
+        if operation_mode == STATE_COMFORT:
```
Same as above.

MartinHjelmare commented on homeassistant/components/fritzhome.py:
```diff
@@ -0,0 +1,60 @@
+    for domain in SUPPORTED_DOMAINS:
+        discovery.load_platform(hass, domain, DOMAIN, {}, config)
+
+    _LOGGER.info('Connected to fritzbox')
```
You can remove this as it's already logged in the core when a component setup is successful.

MartinHjelmare commented on homeassistant/components/switch/fritzhome.py:
```diff
@@ -0,0 +1,66 @@
+http://home-assistant.io/components/switch.fritzhome/
+"""
+import logging
+from homeassistant.components.fritzhome import DOMAIN
```
See comment at the top how to import this domain.

Please add a blank line between import groups: standard python, 3rd party libraries and homeassistant imports.
https://www.python.org/dev/peps/pep-0008/#imports

MartinHjelmare commented on homeassistant/components/switch/fritzhome.py:
```diff
@@ -0,0 +1,66 @@
+class FritzhomeSwitch(SwitchDevice):
+    """The switch class for Fritzhome switches."""
+
+    def __init__(self, hass, device):
```
Don't pass in `hass`. It will be set on the entity when the entity has been added to home assistant. This means you can access `hass` with `self.hass`, after the entity has been added to home assistant.

hthiery committed:
change checking of kwargs to canonical way

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
remove unused self._state

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
Don't overwrite the platform domain

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
Remove parenthesis from import without line break

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
Do not pass hass to the components on init

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
Remove check for available in current_operation

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
Remove redundant logging message

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
Add blank line between standard and hass imports

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
Use states from base climate component

Also add the new state STATE_MANUAL to the base.

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery commented:
What is the best way to detect and reconnect in case of a connection error with a component/platform device? In my case the update function can fail with a HTTPError exception when the session isnt valid any more. Then it is neccessary to do a device reconnect that was done in the platform setup.

MartinHjelmare commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,132 @@
+    ATTR_OPERATION_MODE, ClimateDevice, STATE_ECO, STATE_HEAT, STATE_MANUAL,
+    SUPPORT_OPERATION_MODE, SUPPORT_TARGET_TEMPERATURE)
+from homeassistant.const import PRECISION_HALVES
+from homeassistant.const import TEMP_CELSIUS, ATTR_TEMPERATURE
```
Please group the imports from the same module.
```py
from homeassistant.const import (
    ATTR_TEMPERATURE, PRECISION_HALVES, TEMP_CELSIUS)
```

MartinHjelmare reviewed (changes requested):
Small comment left.

MartinHjelmare commented on homeassistant/components/fritzhome.py:
```diff
@@ -0,0 +1,58 @@
+    try:
+        fritz = Fritzhome(host=host, user=username, password=password)
+        fritz.login()
+        hass.data[DOMAIN] = fritz.get_devices()
```
If you want to be able to reconnect from a platform, you should store the `fritz` instance in `hass.data[DOMAIN]` instead of the devices. Then you can both reconnect and get devices from the platforms.

MartinHjelmare replied:
Pass the `fritz` instance to the device class at init, so you have access to that inside each entity.

hthiery replied:
Each device has a reference to the fritz instance. So it is possible to do a reconnect with that.

MartinHjelmare replied:
I would prefer if you store the `Fritzhome` instance in `hass.data` and then pass that to each entity, instead of accessing a protected attribute of the fritz device.

hthiery replied:
changed according your comment

MartinHjelmare commented on homeassistant/components/climate/fritzhome.py:
```diff
@@ -0,0 +1,132 @@
+
+    def update(self):
+        """Update the data from the thermostat."""
+        self._device.update()
```
To reconnect, add `try... except` around this call, and call the reconnect method in the `except` block.

hthiery replied:
I found a way to do a reconnect. Just added the code and pushed.

hthiery committed:
add reconnect when access failed

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
add device specific attributes

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
group the imports from the same module

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
change domain data to fritz instance

This let us use the fritz instance to reconnect from platform without accessing
protected attributes.

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

MartinHjelmare commented:
What's the difference between these fritz devices and the ones in `homeassistant/components/switch/fritzdect.py`? It's a bit weird that we create a new component and platforms for what seems to be very similar brand of devices.

Would it be possible to build a fritz component with platforms for all fritz devices and depend on the same api library?

hthiery commented:
You are right .. this is pretty the same. The 'old' fritzdect switch device uses an API that is nolonger developed. That was my reason to have another one that will support more new devices. But you're right there should be a way to replace one of both. I dont know what the best way would be to change that.

hthiery committed:
fix typo

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

MartinHjelmare commented:
@BastianPoe you are also working with fritz devices. What do you think about integrating all fritz devices under one component with platforms? Is it possible?

hthiery commented:
Any ideas (@BastianPoe and @MartinHjelmare) how to proceed? Is it a valid aproach to rename the new fritzhome platform/component to fritzdect and use the 'new' API in the existing fritzdect switch component. So there is only one API that has to be maintained?

MartinHjelmare commented:
I definitely think we should use one API library for this and the newest and most maintained is the way to go.

Regarding naming of component:
Whats the relationship with the other fritz box platforms for presence and system monitor. Could they also be included in the same component. They seem to be of the same brand.

Maybe we should call the new component simply `fritz`?

hthiery commented:
All components you mentionend (callmonitor, presence and the dect) are related to the AVM fritzbox device. This is a famous dsl/cable router mainly (AFAIK) known in germany.

You're absolutly right that it would be the best for the user perspective to have one platform that supports all of these components. But from my point of view (as maintainer of the fretzhome API) it is not that easy to bring all interfaces (presence, monitor and dect/smarthome) to one API.

I think in the first step we should include the existing fritzdect and my proposed fritzhome to one maybe called `fritzbox`.

Is it allowed (from homeassistant implementation) to have serveral APIs used in one platform?
E.g. we start a new platform called fritzbox and move all other components to this central point?

MartinHjelmare commented:
I think we can do the transition step wise as you suggest. I don't think we should have more than one API library for the same component. So all platforms that use the new `fritzbox` component should use the same library. Otherwise there's very little point in bringing them under the same component.

I'd like to hear from another member before we go ahead, and preferably also from someone involved in the `fritzdect` switch platform.

BastianPoe commented:
Just had a look into this. The Fritz!Box router exposes the functionality used by HASS via different protocols:
* Home Automation: HTTP API (documentation: https://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/AHA-HTTP-Interface.pdf)
* List of network devices (aka Presence Detection: TR-064 (documentation: https://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/wlanconfigSCPD.pdf)
* Network Monitor
* Call Monitor

I was unable to find a (documented) interface of the Fritz!Box that exposes all functionality used by HASS. Also, I was unable to find a python library that abstracts away all the different interfaces. Currently, there seem to be three libraries in use:
* fritzconnection (network devices, network monitor, call monitor)
* fritzhome (home automation)
* pyfritzhome (home automation)

Hence I believe, that unifying all of HASS' interfaces into one library would first require creating such a unified library.

MartinHjelmare commented:
@BastianPoe I think we don't need to unify all platforms right now, but it would still be better to unify all that we can, using one of the existing libraries.  I think this sounds like a good approach:
> I think in the first step we should include the existing fritzdect and my proposed fritzhome to one maybe called fritzbox.

hthiery commented:
So as i already mentioned my suggestion is to unify the components (switch/fritzdect, switch/fritzhome, climate/fritzhome) to use the python-fritzhome library.

The only risk I see is that renaming/moving the switch/fritzdect component could break something already working in the fritzdect.

As part of that I could also rename the platform/component to fritzbox.

@BastianPoe, @MartinHjelmare  Any suggestions from your side?

MartinHjelmare commented:
@hthiery I think your suggestion sounds good.

@balloob do you have an opinion about renaming some of the existing fritz* platforms and migrating to another interface library?

balloob commented:
We can do a breaking change for the rename. Should be a one-time nuisance.

hthiery committed:
rename platform from fritzhome to fritzbox

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
Add device_state_attributes

Add attributes to have compatiblity to fritzdect.

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
add support for multiple fritzboxes

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
fix pylint issues

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
fixed pyfritzhome version

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
fix import

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
fix component name in requirements_all.txt

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery commented:
@BastianPoe, @MartinHjelmare: I renamed the component/platform to fritzbox and added (hopefully all) features like device_state_attributes from switch/fritzdect to the switch/fritzbox component/platform.

What is the best practice for replacing the fritzdect with the new one? Should we first integrate the new component/platform in the next release and mark the switch/fritzdect as deprecated until the after next?

MartinHjelmare commented:
Sounds good with a deprecation notice during at least one release cycle.

hthiery commented:
@MartinHjelmare: A review with requesting changes seems to be open. I cannot figure out what to do at this point. Could you please clarify.

MartinHjelmare commented:
If you think this is ready, just wait until someone has time to look at your latest changes.

hthiery committed:
upgrade pyfritzhome to 0.3.7

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

c727 commented:
- you have to rename the files in `.coveragerc`
- DEFAULT_HOST is not assigned
- why do we need the `devices:` subkey?
- update docs PR

hthiery committed:
rename platform/component also in .coveragerc

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
use DEFAULT_HOST when no host is in dict

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery commented:
@c727: 
` devices:` is needed to support several fritzbox devices. Or do you know a better way to have a list of hosts with their password and username?

docs PR should be updated. I just found one wrong name.

hthiery commented:
The build failed and it looks to me like a timeout issue of the tests. Is there a way to rebuild the travis job without pushing changed code?

MartinHjelmare commented:
Close and reopen the PR will restart the build.

hthiery commented:
I dont know the process, but the labels should also be renamed from `platform: fritzhome` to `platform: fritzbox` and `platform: climate.fritzhome`to `platform: climate.fritzbox` and `platform: climate.fritzhome` to `platform: climate.fritzbox`.

MartinHjelmare commented on homeassistant/components/fritzbox.py:
```diff
@@ -0,0 +1,74 @@
+                _LOGGER.warning("Configuration error")
+                continue
+
+    hass.data[DOMAIN] = fritz_list
```
Check that this list is not empty before putting it in `hass.data`. If it is empty return `False` so that we don't set up a component with no functioning devices.

MartinHjelmare commented on homeassistant/components/switch/fritzbox.py:
```diff
@@ -0,0 +1,109 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_CURRENT_CONSUMPTION = 'current_consumption'
```
There is already a standard attribute in the switch base class corresponding to this attribute here. We could make that change now, since switching platforms will be a breaking change anyway.

MartinHjelmare commented on homeassistant/components/switch/fritzbox.py:
```diff
@@ -0,0 +1,109 @@
+class FritzboxSwitch(SwitchDevice):
+    """The switch class for Fritzbox switches."""
+
+    def __init__(self, hass, device, fritz):
```
You don't need to pass in `hass`. See below.

MartinHjelmare commented on homeassistant/components/switch/fritzbox.py:
```diff
@@ -0,0 +1,109 @@
+
+    def __init__(self, hass, device, fritz):
+        """Initialize the switch."""
+        self.units = hass.config.units
```
The entity will have access to `hass` when it has been added to home assistant, so then you can use `self.hass.config.units`, so you can remove this attribute here.

MartinHjelmare commented on homeassistant/components/switch/fritzbox.py:
```diff
@@ -0,0 +1,109 @@
+
+        if self._device.has_temperature_sensor:
+            attrs[ATTR_TEMPERATURE] = "{}".format(
+                self.units.temperature(self._device.temperature, TEMP_CELSIUS))
```
Use `self.hass.config.units.temperature(...)`

MartinHjelmare commented on homeassistant/components/switch/fritzbox.py:
```diff
@@ -0,0 +1,109 @@
+            attrs[ATTR_TEMPERATURE] = "{}".format(
+                self.units.temperature(self._device.temperature, TEMP_CELSIUS))
+            attrs[ATTR_TEMPERATURE_UNIT] = "{}".format(
+                self.units.temperature_unit)
```
See above.

MartinHjelmare commented on homeassistant/components/climate/fritzbox.py:
```diff
@@ -0,0 +1,155 @@
+    def device_state_attributes(self):
+        """Return the device specific state attributes."""
+        attrs = {
+            ATTR_STATE_DEVICE_LOCKED: self._device.device_lock,
```
What does it mean that the device is locked?

hthiery replied:
The buttons on the device are locked. Locking can be enabled either on the device or with the UI/API.

I dont know if these values can be used in automation but I just added these values to the attributes.

hthiery replied:
Is this something that isn't common? I don't find much documented device specifc attributes.

MartinHjelmare replied:
Good idea to add it to the docs. :+1:

MartinHjelmare reviewed (dismissed):
Looks good!

Currently nothing happens when the user selects manual mode for operation mode. I think you should remove that mode from the list of possible selectable modes, but still return it in current operation, when the user overrides a temperature.

See below for more comments.

MartinHjelmare commented on homeassistant/components/climate/fritzbox.py:
```diff
@@ -0,0 +1,155 @@
+
+SUPPORT_FLAGS = (SUPPORT_TARGET_TEMPERATURE | SUPPORT_OPERATION_MODE)
+
+OPERATION_LIST = [STATE_HEAT, STATE_ECO, STATE_MANUAL]
```
Remove `STATE_MANUAL` from this list?

hthiery committed:
add config schema for dict

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
remove check

The check since since the config scheme takes case.

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
add check for empty devices

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
use standard attribute from base class

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
remove STATE_MANUAL from operation list

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
remove set DEFAULT_HOST

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
don't pass hass to the SwitchDevice

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
remove unsed DEFAULT_HOST

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
refactored device attributes

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
add info output if no fritzbox is configured

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery commented:
Should I increase `ha_release` field in the documentation header or is this done automatically?

bakedraccoon commented:
You should add it

MartinHjelmare commented on homeassistant/components/switch/fritzbox.py:
```diff
@@ -0,0 +1,109 @@
+                                                   TEMP_CELSIUS))
+            attrs[ATTR_TEMPERATURE_UNIT] = "{}".format(
+                self.hass.config.units.temperature_unit)
+            attrs[ATTR_STATE_DEVICE_LOCKED] = self._device.device_lock
```
This is a copy from above.

MartinHjelmare commented on homeassistant/components/switch/fritzbox.py:
```diff
@@ -0,0 +1,109 @@
+            attrs[ATTR_STATE_LOCKED] = self._device.lock
+
+        if self._device.has_temperature_sensor:
+            attrs[ATTR_TEMPERATURE] = "{}".format(
```
If you don't need to format the string, just do `str(value)`.

MartinHjelmare commented on homeassistant/components/fritzbox.py:
```diff
@@ -0,0 +1,82 @@
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_DEVICES, default=[]):
```
Since this is required, you can remove the default value.

MartinHjelmare commented on homeassistant/components/fritzbox.py:
```diff
@@ -0,0 +1,82 @@
+
+    configured_devices = config[DOMAIN].get(CONF_DEVICES)
+    for device in configured_devices:
+        try:
```
Please only wrap the lines with `try... except` that require that. I guess that's `fritzbox.login()`.

MartinHjelmare reviewed (commented):
Looks good! Some small comments left to fix.

MartinHjelmare commented on homeassistant/components/climate/__init__.py:
```diff
@@ -42,6 +42,7 @@
 STATE_COOL = 'cool'
 STATE_IDLE = 'idle'
 STATE_AUTO = 'auto'
+STATE_MANUAL = 'manual'
```
I just remembered that @balloob thought that this should be a hold mode instead. See:
https://github.com/home-assistant/home-assistant/pull/11454#discussion_r171114275

hthiery replied:
Does hold mode means that this temperature value should be hold? In the fritzbox implementation this will only work until the schedule configured in the fritzbox timesheet will switch again. e.g. 22:00 to  eco mode.

MartinHjelmare replied:
Yeah. Ok, so then in the fritzbox case it doesn't sound as a hold mode feature since it doesn't override the schedule. So then this manual mode should still be operational mode? @balloob?

MartinHjelmare commented on homeassistant/components/climate/fritzbox.py:
```diff
@@ -0,0 +1,153 @@
+            return STATE_HEAT
+        elif self._target_temperature == self._eco_temperature:
+            return STATE_ECO
+        return STATE_MANUAL
```
Then I don't know how we should handle this case. What should be reported here then?

hthiery commented on homeassistant/components/switch/fritzbox.py:
```diff
@@ -0,0 +1,109 @@
+            attrs[ATTR_TOTAL_CONSUMPTION_UNIT] = "{}".format(
+                ATTR_TOTAL_CONSUMPTION_UNIT_VALUE)
+
+        if self._device.has_switch:
```
yes .. is checked above

hthiery committed:
small fixes according review comment

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
remove unneeded default value

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
remove non required code from try..except block

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
Merge remote-tracking branch 'upstream/dev' into fritzhome

hthiery committed:
line break for line that is too long

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

hthiery committed:
remove too many empty lines

Signed-off-by: Heiko Thiery <heiko.thiery@gmail.com>

Vetti52 commented:
mybe you can explain:
I want to control a powerline 546e over homeassistant. Will this work?? 
After installation, everything looks fine, but it won't start.
My entry in configuration.yaml:

`fritzbox:`
   `devices:`
    `- host: 192.168.1.1`
      `username: USER`
      `password: PASS`

I tried "renaming" to fritzhome, no success.
What else to do?

my first checks did not give any hints:

`(homeassistant) homeassistant@raspberrypi:/srv/homeassistant $ pip3 show pyfritzhome`
`Name: pyfritzhome`
`Version: 0.3.7`
`Summary: Fritz!Box Smarthome Python Library`
`Home-page: http://github.com/hthiery/python-fritzhome`
`Author: Heiko Thiery`
`Author-email: heiko.thiery@gmail.com`
`License: LGPLv2+`
`Location: /srv/homeassistant/lib/python3.5/site-packages`
`Requires: requests`

and 

`(homeassistant) homeassistant@raspberrypi:/srv/homeassistant $ hass --script check_config`
`Testing configuration at /home/homeassistant/.homeassistant`
`Failed config`
`  General Errors:` 
`    - Component not found: fritzbox`
`  `
`Successful config (partial)`

hthiery commented:
> I want to control a powerline 546e over homeassistant. Will this work??

Yet I have no experience with that device.

> After installation, everything looks fine, but it won't start.
> My entry in configuration.yaml:
> 
> fritzbox:
> devices:
> - host: 192.168.1.1
> username: USER
> password: PASS
> 
> I tried "renaming" to fritzhome, no success.

How did you test that code? Copied the files from that pull-request to your installation? You can also try my custom_compoent version from https://github.com/hthiery/homeassistant.

> What else to do?

Please open an issue on https://github.com/hthiery/python-fritzhome and let's do further discussion there.

Vetti52 commented:
Ok, I did copy the complete custom_component version into the srv directory, as described there.
Now, it works!!!
FYI, the Plug can be switched. However, the info returned, is still incomplete:
`Powerline546E`
`19 minutes ago`
`current consumption unit     W`
`total consumption unit         kWh`
`current consumption           0.0`
`total consumption               6.968`
compared to the reported data, the total consumption is that per day. There are also data per month and per year.
Anyway, it is running
Thanks!

hthiery commented:
that sounds good ... please open an issue under the custom_component github repo to do further investigations.

also it would be interesting to have an output of the python-fritzhome cli tool with some debug information/traces.

hthiery commented:
Is there a chance to get this into one of the next releases?

MartinHjelmare commented:
I suggest we see what comes out of https://github.com/home-assistant/architecture/issues/22 first so we know how to handle the climate modes.

Please get involved in that issue if you have input that isn't covered yet.

escoand commented:
I've just pull-requested the auto-discovery to the AVM FRITZ devices: https://github.com/home-assistant/netdisco/pull/186
Therefore it would be really nice if you could add this to the new component.
So everything could run out-of-the-box when released.

The discovered information from netdisco modules looks with the example_service.py like that:
```
2018-04-06 19:54:29.952638 - Found new service: fritzbox {'host': '192.168.1.1', 'port': 49000, 'ssdp_description': 'http://192.168.1.1:49000/fboxdesc.xml', 'name': 'FRITZ!Box 7490 (UI)', 'model_name': 'FRITZ!Box 7490 (UI)', 'model_number': '1und1', 'serial': None, 'manufacturer': 'AVM Berlin', 'udn': 'uuid:123402409-bccb-40e7-8e6c-C80E14737DA2'}
2018-04-06 19:54:29.952771 - Found new service: fritzbox {'host': '192.168.1.39', 'port': 49000, 'ssdp_description': 'http://192.168.1.39:49000/fboxdesc.xml', 'name': 'FRITZ!Powerline 546E', 'model_name': 'FRITZ!Powerline 546E', 'model_number': 'avm', 'serial': None, 'manufacturer': 'AVM Berlin', 'udn': 'uuid:123402409-bccb-40e7-8e6c-5C4979F18987'}
```
For you the only interesting information should be `host` and maybe `name` to get the correct type of component.

hthiery commented:
I dont understand what we could achieve with the discovery since to get the fritzbox smarthome components into HA we need an additional fritzbox user with the right permissions.

escoand commented:
Hmmm, I have to admit you're right.

Is there some kind of login screen already implemented for other components?
Would be really helpful to develop the discovery in general in this direction.

escoand commented:
The Plex component has something similar: https://www.home-assistant.io/components/media_player.plex/

MartinHjelmare commented:
Config entries allow user confirmation upon discovery. But that's for another PR.

You can look at the hue component if you're interested in config entries.

pvizeli committed:
Merge branch 'dev' into fritzhome

pvizeli reviewed (approved):
Let's merge it and do others in follows PRs. Thanks for hardwork of all participants
